### PR TITLE
Lazy parallelization of specification parsing. #1364

### PR DESCRIPTION
**Pull request by BugDiver**

### CODE REVIEW

File: parser/specFileCollection.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
+package parser
```
Comment (sriv) [inline parser/specFileCollection.go]: need licence header

File: parser/specFileCollection_test.go [hunk 2]
```diff
@@ -0,0 +1,21 @@
+package parser
```
Comment (sriv) [inline parser/specFileCollection_test.go]: license header

File: util/rlimitUtil_windows.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+
+// Gauge is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Gauge.  If not, see <http://www.gnu.org/licenses/>.
+
+// +build windows
```
Comment (sriv) [inline util/rlimitUtil_windows.go]: let's follow one convention. `_windows` and the `+build` are two different ways of making platform specific implementation. I guess either one is enough

File: parser/parse.go [hunk 4]
```diff
@@ -54,29 +55,56 @@ import (
...
+
 func ParseSpecFiles(specFiles []string, conceptDictionary *gauge.ConceptDictionary, buildErrors *gauge.BuildErrors) ([]*gauge.Specification, []*ParseResult) {
-	parseResultsChan := make(chan *ParseResult, len(specFiles))
-	specsChan := make(chan *gauge.Specification, len(specFiles))
+	sfc := NewSpecFileCollection(specFiles)
+	piChan := make(chan *parseInfo)
+	limit := len(specFiles)
+	rLimit, e := util.RLimit()
+	if e == nil && rLimit < limit {
+		logger.Infof(true, "No of specifcations %d is higher than Max no of open file descriptors %d.\n"+
```
Comment (sriv) [inline parser/parse.go]: this could be a debug level log.

File: parser/specFileCollection.go [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+)
+
+type SpecFileCollection struct {
+	mutex     *sync.Mutex
+	index     int
+	specFiles []string
+}
+
+func NewSpecFileCollection(s []string) *SpecFileCollection {
+	return &SpecFileCollection{specFiles: s, index: 0, mutex: &sync.Mutex{}}
```
Comment (sriv) [inline parser/specFileCollection.go]: the `specFiles` is not mutated as far as I see. What is the purpose of the mutex?

Comment (BugDiver) [reply parser/specFileCollection.go]: The `specFiles` is not mutated but the `index` is mutated. The `mutex` is there to ensure that two goroutines are not parsing the same file.

Comment (sriv) [reply parser/specFileCollection.go]: Are `SpecFileCollection` and `NewSpecFileCollection` used outside `parser` package? I could not locate any external usage. If not, then these need not be exported.

**Pull request merged**